Updated CockroachDB example

haaarshsingh opened the pull request:
In response to #3690 

Now that we have Prisma migrate available with CockroachDB, the example given in this repository is out-of-date. I've updated the schema, script, tests, and the instructions of the `README` in accordance to the example given over in `sql-server`.

Changes:

 1. Update `schema.prisma` in accordance with the example given on `sql-server`, adding the `Comments` model.
 2. Remove the `dbinit` + `schema-introspected.prisma` as we're using `prisma migrate`.
 3. Updated `@types/node` to the latest version.
 4. Updated `script.ts` to match the example given in `sql-server`, as we've added the `Comments` model.
 5. Update the Jest tests as we've updated the `Comments` model 
 6. Changed information regarding setup on README

haaarshsingh committed:
Updated relations to match SQL server example

haaarshsingh committed:
Removed introspected schema + SQL file

haaarshsingh committed:
Removed introspected schema + dbinit

haaarshsingh committed:
Update @types/node version on example

haaarshsingh committed:
Updated script in accordance to sql-server example

haaarshsingh committed:
Updated tests in accordance to sql-server

haaarshsingh committed:
Changed README instructions

haaarshsingh committed:
Update the example overview

haaarshsingh reviewed (commented):
Here, the CockroachDB test fails as there is no `dbinit.sql` file in the project in the example now makes use of Prisma migrate instead. I believe this might either be an issue with the GitHub action or the docker file? I'm unsure, I'll take a look.

haaarshsingh committed:
GitHub worflow for CockroachDB now uses migrate

haaarshsingh committed:
Update CockroachDB Serverless instructions on README

haaarshsingh reviewed (commented):
I'm unsure why [test (typescript/graphql-nestjs-sdl-first)](https://github.com/prisma/prisma-examples/runs/6064475749?check_suite_focus=true#logs) is failing - I've looked into it, and it doesn't seem to be related to anything that I've changed.

haaarshsingh committed:
Removed creating CockroachDB instances from the workflow

Looking at other examples in this repository, it seems as if we don't need to create the CockroachDB instance as we're not using the CRDB client anywhere.

haaarshsingh commented:
I'll review the workflow file again - but this might have something to do with Docker? I'm unsure.

Docker compose is working for me. Unsure why the instance isn't spinning up in thet test.

![image](https://user-images.githubusercontent.com/69592270/163906124-af30eb82-d735-45db-8803-b50b682e76cf.png)

haaarshsingh commented:
My mistake, I just realised that I don't run `prisma migrate dev` on the GitHub workflow.

haaarshsingh committed:
Update workflow with to use prisma migrate

haaarshsingh committed:
Merge remote-tracking branch 'origin/cockroach-fix' into cockroach-fix

janpio commented on .github/workflows/test-cockroachdb.yaml:
```diff
@@ -27,20 +27,9 @@ jobs:
 
     steps:
       - uses: actions/checkout@v3
-      - name: Start a single CockroachDB instance with Docker
-        env:
-          COCKROACHDB_DOCKER_TAG: 'cockroachdb/cockroach:latest-v21.2'
-        run: |
-          docker pull $COCKROACHDB_DOCKER_TAG
-          docker run -d --name roach --hostname roach -p 26257:26257 -p 8080:8080 -v "${{ github.workspace }}:/app" $COCKROACHDB_DOCKER_TAG start-single-node --insecure
-          sudo apt update && sudo apt install wait-for-it -y
-          wait-for-it -h localhost -p 26257
```
I think this part (lines 30 to 37) is still needed so a CockroachDB instance is actually started that we can later connect to.

haaarshsingh replied:
Oh I see, interesting. There wasn't a similar line to spin up the SQL server instance on the other workflow, so I removed it. I'll add it back and see how it plays out.

janpio replied:
SQL Server might already be installed and running on the GitHub Actions container we are using.

haaarshsingh replied:
I believe I've fixed all issues related to the actions and the tests now. The tests run successfully on my machine, and the workflow ran successfully on the [670c541](https://github.com/prisma/prisma-examples/pull/3691/commits/670c54161443557c259548e83cf3b8f92499dec3)

haaarshsingh committed:
Start a CRDB instance with Docker in workflow

haaarshsingh committed:
Change order of actions

Before, actions/checkout@v3 was happening before starting the CockroachDB instance with Docker.

haaarshsingh committed:
Changed expected type on tags in Prisma test

haaarshsingh commented:
Correct me if I'm wrong, @janpio I believethis failing workflow has nothing to do with what I've changed?

janpio commented:
Indeed, I restarted and now it passes🍾

ruheni commented on databases/cockroachdb/README.md:
```diff
@@ -1,11 +1,8 @@
 # CockroachDB example
 
-This example shows how to:
+This example shows how to connect Prisma to a CockroachDB database, create the database schema with [Prisma Migrate](https://www.prisma.io/migrate) and use [Prisma Client](https://www.prisma.io/client) in a **Node.js script** to read and write data in a CockroachDB database.
```
Let's use the bullet points to list what the example does – it would make it easier to read.

ruheni commented on databases/cockroachdb/README.md:
```diff
@@ -1,11 +1,8 @@
-- Read and write data to the database using [Prisma Client](https://www.prisma.io/client)
+You can find the Prisma schema in [`./prisma/schema.prisma`](./prisma/schema.prisma) from which the SQL for the database schema is generated.
 
 > **Note:** CockroachDB support in Prisma is currently in [**Preview**](https://www.prisma.io/docs/about/prisma/releases#preview), and **Prisma Migrate isn't supported for now**. For this reason, you will create the database schema with raw SQL.
```
Remove this note

ruheni commented on databases/cockroachdb/README.md:
```diff
@@ -155,105 +115,25 @@ DATABASE_URL="postgresql://root@localhost:26257/prisma?sslmode=disable"
 
 > **Note:** If you're using CockroachDB Serverless, see [`.env.example`](./.env.example) for more information on how `DATABASE_URL` should look like with the cluster configruation.
```
It seems this file is missing – could you include it in the PR?

haaarshsingh replied:
This file isn't in the main repository, right?

![gitignore](https://user-images.githubusercontent.com/69592270/165003168-7c084554-53f1-4c1a-9f7c-8c93b5c62ab9.png)

I'll also have to edit the `.gitignore` for this, if that's okay.

ruheni replied:
Sounds good. 👍🏾

ruheni commented on databases/cockroachdb/README.md:
````diff
@@ -155,105 +115,25 @@ DATABASE_URL="postgresql://root@localhost:26257/prisma?sslmode=disable"
-
-```sh
-npx prisma generate
+npx prisma migrate dev --name "init"
````
```suggestion
npx prisma migrate dev --name init
```

ruheni commented on databases/cockroachdb/prisma/schema.prisma:
```diff
@@ -7,3 +7,40 @@ datasource db {
+model Tag {
+  id    String @id @default(uuid())
+  tag   String @unique
+  posts Post[] @relation("TagToPost")
```
```suggestion
  posts Post[]
```
Let's use Prisma's implicit m-n

ruheni commented on databases/cockroachdb/prisma/schema.prisma:
```diff
@@ -7,3 +7,40 @@ datasource db {
+  authorId  String
+  author    User      @relation(fields: [authorId], references: [id])
+  comments  Comment[]
+  tags      Tag[]     @relation("TagToPost")
```
```suggestion
  tags      Tag[]  
```
Let's use Prisma's implicit m-n

ruheni commented on databases/cockroachdb/package.json:
```diff
@@ -16,7 +16,7 @@
   "devDependencies": {
     "@types/jest": "27.4.1",
-    "@types/node": "17.0.14",
+    "@types/node": "^17.0.24",
```
```suggestion
    "@types/node": "17.0.24",
```

ruheni reviewed (changes requested):
Thanks for the contribution  @harshhhdev 🙏🏽

Just a few minor adjustments and then we'll merge the PR.

haaarshsingh committed:
Added prisma migrate init suggestion

Co-authored-by: Alex Ruheni <33921841+ruheni@users.noreply.github.com>

haaarshsingh committed:
Use Prisma's implicit m-n (suggestion)

Co-authored-by: Alex Ruheni <33921841+ruheni@users.noreply.github.com>

haaarshsingh committed:
Fix @types/node version in package.json (suggestion)

Co-authored-by: Alex Ruheni <33921841+ruheni@users.noreply.github.com>

haaarshsingh committed:
Use Prisma's implicit m-n (suggestion)

Co-authored-by: Alex Ruheni <33921841+ruheni@users.noreply.github.com>

haaarshsingh committed:
Merge branch 'prisma:latest' into cockroach-fix

haaarshsingh committed:
Fix README issues

haaarshsingh committed:
Added connection string example for CRDB serverless

haaarshsingh commented:
Hey @ruheni, thanks for getting back to me and reviewing this pull request! I've updated all changes according to how you wanted them, if there's anything else you'd like me to fix/add then please do let me know :tada:

janpio commented:
```
FAIL tests/prisma.test.ts
  example test with Prisma Client
    ✓ test query (6 ms)
    ✓ create tags (42 ms)
    ✕ create two user with posts comments and tags (60 ms)
    ✓ Get all published posts with a given tag (9 ms)
    ✓ Create unpublished post with tags and comments for an existing user and then publish (29 ms)
    ✓ Find comments by a user (6 ms)
    ✓ Fetch posts by a user (4 ms)

  ● example test with Prisma Client › create two user with posts comments and tags

    TypeError: Cannot read property 'id' of undefined

      [17](https://github.com/prisma/prisma-examples/runs/6152374903?check_suite_focus=true#step:7:17)3 |     expect(user2.posts.length).toEqual(3)
      174 |     expect(user2.posts[0].authorId).toEqual(user2.id)
    > 175 |     expect(user2.posts[0].tags[0].id).toBeTruthy()
          |                                   ^
      176 |     expect(
      177 |       user2.comments[0].comment.toLowerCase().includes('thanks'),
      178 |     ).toBeTruthy()

      at Object.<anonymous> (tests/prisma.test.ts:175:[35](https://github.com/prisma/prisma-examples/runs/6152374903?check_suite_focus=true#step:7:35))

Test Suites: 1 failed, 1 total
Tests:       1 failed, 6 passed, 7 total
```

haaarshsingh commented:
Huh, this is definitely strange. Before the workflow wasn't running properly because the order of the posts created were incorrect, and now it's not running again for the same reason? I'll update the arrays to what they were previously and see if it's any better.

ruheni commented:
Seems the test was flaky – I re-run the tests and it seems everything is fine 🙂

I'll go ahead and merge this – if it fails again, I'll look into the test
